src/services/impl/ThreadPool.cpp

```diff
@@ -0,0 +1,82 @@
+#include "ThreadPool.hh"
+
+#include "RA_Log.h"
+
+namespace ra {
+namespace services {
+namespace impl {
+
+ThreadPool::~ThreadPool()
+{
+    Shutdown(true);
+}
+
+void ThreadPool::Initialize(size_t nThreads) noexcept
+{
+    assert(m_vThreads.empty());
+
+    RA_LOG_INFO("Initializing %zu worker threads", nThreads);
+
+    for (size_t i = 0; i < nThreads; ++i)
+        m_vThreads.emplace_back(&ThreadPool::RunThread, this);
+}
+
+void ThreadPool::RunThread() noexcept
+{
+    do
+    {
+        // check for work
+        do
+        {
+            std::function<void()> pNext;
+            {
+                std::unique_lock<std::mutex> lock(m_oMutex);
```

Comment:
There isn't a timeout because this is specifically designed to wait until notified. `scoped_lock` doesn't support `condition` variables (`m_cvMutex`) which are used to pop out of the lock.

Reply:
`std::condition_variable_any` supports any `scoped_lock`.